partner_changeset/models/res_partner_changeset.py

```diff
@@ -0,0 +1,516 @@
+
+    @api.model
+    def get_field_for_type(self, field, prefix):
+        assert prefix in ('origin', 'old', 'new')
+        field_type = self._type_to_suffix.get(field.ttype)
+        if not field_type:
+            raise NotImplementedError(
+                'field type %s is not supported' % field_type
+            )
+        return '%s_value_%s' % (prefix, field_type)
+
+    @api.multi
+    def get_origin_value(self):
+        self.ensure_one()
+        field_name = self.get_field_for_type(self.field_id, 'origin')
+        return self[field_name]
+
+    @api.multi
+    def get_new_value(self):
+        self.ensure_one()
+        field_name = self.get_field_for_type(self.field_id, 'new')
+        return self[field_name]
+
+    @api.multi
+    def set_old_value(self):
+        """ Copy the value of the partner to the 'old' field """
+        for change in self:
+            # copy the existing partner's value for the history
+            old_value_for_write = self._value_for_changeset(
+                change.changeset_id.partner_id,
+                change.field_id.name
+            )
+            old_field_name = self.get_field_for_type(change.field_id, 'old')
+            change.write({old_field_name: old_value_for_write})
+
+    @api.multi
+    def apply(self):
+        """ Apply the change on the changeset's partner
+
+        It is optimized thus that it makes only one write on the partner
+        per changeset if many changes are applied at once.
+        """
+        changes_ok = self.browse()
+        key = attrgetter('changeset_id')
+        for changeset, changes in groupby(self.sorted(key=key), key=key):
+            values = {}
+            partner = changeset.partner_id
+            for change in changes:
+                if change.state in ('cancel', 'done'):
+                    continue
+
+                field = change.field_id
+                value_for_write = change._convert_value_for_write(
+                    change.get_new_value(),
+                    partner
+                )
+                values[field.name] = value_for_write
+
+                change.set_old_value()
+
+                changes_ok |= change
+
+            if not values:
+                continue
+
+            previous_changesets = self.env['res.partner.changeset'].search(
+                [('date', '<', changeset.date),
+                 ('state', '=', 'draft'),
+                 ('partner_id', '=', changeset.partner_id.id),
+                 ],
+                limit=1,
+            )
+            if previous_changesets:
+                raise exceptions.Warning(
+                    _('This change cannot be applied because a previous '
+                      'changeset for the same partner is pending.\n'
+                      'Apply all the anterior changesets before applying '
+                      'this one.')
+                )
+
+            partner.with_context(__no_changeset=True).write(values)
```

Comment:
I'm not confident in this as you create new cache at each loop iteration.

Maybe a better approach should be a other 'write' method on partner